sympy/functions/elementary/piecewise.py

```diff
@@ -664,3 +993,234 @@ def piecewise_fold(expr):
+    ...     k, v
+    (-1, [(-oo, 1), (2, 4), (5, oo)])
+    (0, [(1, 2)])
+    (1, [(4, 5)])
+
+    There may also be intervals wherein the right boundary is greater
+    than the left:
+
+    >>> for k, v in sorted(replace(x, 6).items()):
+    ...     (k, v)
+    (-1, [(-oo, 2), (2, 4), (5, oo)])
+    (0, [(2, 2)])
+    (1, [(4, 5)])
+    """
+    if not any(i for i in (abi, dict, union)):
+        dict = True
+    if [bool(i) for i in (abi, dict, union)].count(True) != 1:
+        raise ValueError('only 1 of abi, dict and union may be True')
+
+    a, b = A
+    c, d = B
+    oo = S.Infinity
+    # the real line is broken into pieces which are compactly
+    # represented as -oo, i1, H1|L2, i2, H2|L3, i3, ..., oo
+    # where i1, i2, ...., refer to the interval A (0), B (1) or
+    # neither (-1) as covering that interval and Hi|Lj means
+    # either the hi boundary of the ith interval or the lo
+    # boundary of the jth interval
+    if (d <= a) == True:
+        c = Min(c, d)
+        br = (
+        -oo, -1,
+        c, 1,
+        d, -1,
+        a, 0,
+        b, -1,
+        oo)
+    elif (b <= c) == True:
+        a = Min(a, b)
+        br = (
+        -oo, -1,
+        a, 0,
+        b, -1,
+        c, 1,
+        d, -1,
+        oo)
+    else:
+        br = (
+        -oo, -1,
+        Min(a, c), 1,
+        Min(a, d), -1,
+        a, 0,
+        b, -1,
+        Max(b, c), 1,
+        Max(b, d), -1,
+        oo)
+
+    abi = []
+    for _ in range(1, len(br) - 1, 2):
+        a, b, i = br[_ - 1], br[_ + 1], br[_]
+        if a == b:
+            continue
+        abi.append((a, b, i))
+    if not union and not dict:
+        return abi
+    if union:
+        rv = []
+        for a, b, i in abi:
+            if i == -1:
+                continue
+            K = (a, b)
+            if not rv or a != rv[-1][1]:
+                # it's a new interval
+                rv.append(K)
+            else:
+                # it joins with the last one
+                rv[-1] = rv[-1][0], K[1]
+        return rv
+    # dict
+    rv = {}
+    for a, b, i in abi:
+        rv.setdefault(i, []).append((a, b))
+    return rv
+
+
+def _clip(A, B, k):
+    """Return interval B as intervals that are covered by A (keyed
+    to k) and all other intervals of B not covered by A keyed to -1.
+
+    The reference point of each interval is the rhs; if the lhs is
+    greater than the rhs then an interval of zero width interval will
+    result, e.g. (4, 1) is treated like (1, 1).
+
+    Examples
+    ========
+
+    >>> from sympy.functions.elementary.piecewise import _clip
+    >>> _clip((1, 3), (2, 4), 0)
+    [(2, 3, 0), (3, 4, -1)]
+
+    Interpretation: interval portion (2, 3) of interval (2, 4) is
+    covered by interval (1, 3) and is keyed to 0 as requested;
+    interval (3, 4) was not covered by (1, 3) and is keyed to -1.
+    """
+    a, b = B
+    c, d = A
+    A = c, d = Min(Max(c, a), b), Min(Max(d, a), b)
+    p = _pair(A, B, abi=True)
```

Comment:
I wonder if the previous form could have been used. After

    a, b = B
    c, d = A
    B = a, b = Min(a, b), b
    c, d = Min(Max(c, a), b), Min(Max(d, a), b)
    A = c, d = Min(c, d), d

we should have `a <= c <= d <= b`. Then it would suffice to return those of the three pieces `(a, c, -1)`, `(c, d, k)` and `(d, b, -1)` that are not redundant. Or maybe simply `(a, b, -1)` in case `c == d`.